### ISSUE TITLE
FFN function approximator in a multi-dimensional space doesn't converge

### ISSUE THREAD

**Issue by yurivict** [mlpack/mlpack#3477]
#### Issue description

The below testcase trains a FFN on a function in a 5-dimensional space. The function is a set of gaussian functions with different centers in the hyper-quadrant ```[-1,+1]```. Thus 5-dim → 5-dim function is very smooth and well behaved.

However, training doesn't descend too deep in terms of the loss function. The loss only drops from 0.949931 down to 0.325896
Then the ```Predict()``` function always returns the same point regardless of the supplied value point, which is obviously wrong.

#### Your environment

 * version of mlpack: 4.0.1
 * operating system: FreeBSD 13.1
 * compiler: clang-14
 * version of dependencies (Boost/Armadillo): boost-libs-1.81.0, armadillo-12.2.0

#### Steps to reproduce

---compile this testcase.cpp---
```
#include <mlpack.hpp>

#include <array>
#include <cmath>
#include <random>

#include <time.h>

using namespace mlpack;

#define NUM_ARGS 5
#define NUM_VALS 5
#define NUM_SAMPLES 5000


static std::mt19937 rng(::time(nullptr));
static std::uniform_real_distribution<double> rngDouble(0, 1);

typedef std::array<double, NUM_ARGS> PointArg;
typedef std::array<double, NUM_VALS> PointVal;

static double sq(double x) {
	return x*x;
}
static double gauss(PointArg x, PointArg ctr) {
	double sum = 0;

	unsigned i = 0;
	for (auto xi : x)
		sum += sq(xi - ctr[i++]);

	sum /= NUM_ARGS; // normalize

	return std::sqrt(sum);
}

static PointArg randomPoint() {
	PointArg pt;
	for (auto &x : pt)
		x = rngDouble(rng);

	return pt;
}

static PointVal Fn(PointArg arg) {
	return {
		gauss(arg, PointArg{0.0, 0.4, 0.3, 0.2, 0.1}),
		gauss(arg, PointArg{0.1, 0.0, 0.4, 0.3, 0.2}),
		gauss(arg, PointArg{0.2, 0.1, 0.0, 0.4, 0.3}),
		gauss(arg, PointArg{0.3, 0.2, 0.1, 0.0, 0.4}),
		gauss(arg, PointArg{0.4, 0.3, 0.2, 0.1, 0.0})
	};
}

static void printTable(const char *what, const arma::mat &table) {
	std::cout << what << " >>>" << std::endl;

	auto ncols = table.n_cols;
	auto nrows = table.n_rows;

	for (unsigned c = 0; c < ncols; c++) {
		std::cout << " " << (c+1);
		for (unsigned r = 0; r < nrows; r++)
			std::cout << " " << table(r, c);
		std::cout << std::endl;
	}

	std::cout << what << " <<<" << std::endl;
}

int main() {
	// generate train data
	arma::mat trainValues(NUM_ARGS, NUM_SAMPLES);
	arma::mat trainLabels(NUM_VALS, NUM_SAMPLES);
	for (unsigned s = 0; s < NUM_SAMPLES; s++) {
		// random arg
		PointArg arg = randomPoint();

		// label
		PointVal label = Fn(arg);

		// fill data and labels
		for (unsigned r = 0; r < NUM_ARGS; r++)
			trainValues(r, s) = arg[r];
		for (unsigned r = 0; r < NUM_VALS; r++)
			trainLabels(r, s) = label[r];
	}
	printTable("trainValues", trainValues);
	printTable("trainLabels", trainLabels);

	// build the model
	FFN<MeanSquaredError, HeInitialization> model;
	model.Add<Linear>(100);
	model.Add<Sigmoid>();
	model.Add<Linear>(500);
	model.Add<Sigmoid>();
	model.Add<Linear>(500);
	model.Add<Sigmoid>();
	model.Add<Linear>(100);
	model.Add<Sigmoid>();
	model.Add<Linear>(NUM_VALS);

	// train
	model.Train(trainValues, trainLabels, ens::PrintLoss(std::cout));

	// test
	{ // few of existing training points
		for (auto sample : {0, 111, 1023}) {
			arma::mat onePoint(NUM_ARGS, 1);
			arma::mat prediction(NUM_VALS, 1);
			for (unsigned r = 0; r < NUM_ARGS; r++)
				onePoint(r, 0) = trainValues(r, sample);

			model.Predict(onePoint, prediction);

			std::cout << "Test: for sample=" << sample << std::endl;
			printTable("test-sample/values", onePoint);
			printTable("test-sample/prediction", prediction);
		}
	}

	return 0;
}
```

#### Expected behavior

Predictions for the tested training points should match training values.

#### Actual behavior

It always predicts the same value, which is wrong.

**Comment by yurivict** [comment]
The default optimizer fails, but the ```RMSProp``` works.
